### PR TITLE
Friendship v2

### PR DESCRIPTION
**Pull request by martinnajjar12**
### Best practices for Friendship table:

In this milestone, we tried to follow the best practices for creating a friendship table with 2 rows.

**Note:** Please after running `bundle install` run `yarn install --check-files`.

### CODE REVIEW

**Review by Dandush03 — CHANGES_REQUESTED** [review]
# Status: :red_circle: Require Changes :red_circle: 
### Hey Guys! you did a great job on this attempt. but we most fix some issue in order to move forward 

 - [x] **Friendship Model:** You were given [this article](https://explainextended.com/2009/03/07/selecting-friends) to read in order to improve the method which confirms a friend request, check the article if you haven't and read [this issue](https://stackoverflow.com/questions/4219979/sql-best-practice-for-a-friendship-table) in StackOverflow. we should have something similar to this:
```ruby
# Friendship Model

def confirm_friend
  self.update_attributes(confirmed: true)
  Friendship.create!(friend_id: self.user_id,
                  user_id: self.friend_id,
                  confirmed: true)
end
```
  - [x] **User#confirmed_friendships** Thanks to the previews recommendation now we can improve our ways of getting our friend list, like this:
```ruby
# User Model

has_many :confirmed_friendships, -> { where confirmed: true }, class_name: "Friendship"
has_many :friends, through: :confirmed_friendships
```
- [x] **User#pending_friends** can be defined as association with SQL query instead of processing objects in application memory (as the database was created to store and query data it is always more efficient than ruby)
```ruby
# User model

# Users who needs to confirm friendship
has_many :pending_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "user_id"
has_many :pending_friends, through: :pending_friendships, source: :friend
```
- [x] **User#friend_requests** can be defined as association with SQL query instead of processing objects in application memory (as the database was created to store and query data it is always more efficient than ruby)
```ruby
# User model

# Users who requested to be friends (needed for notifications)
has_many :inverted_friendships, -> { where confirmed: false }, class_name: "Friendship", foreign_key: "friend_id"
has_many :friend_requests, through: :inverted_friendships
```
- [x]  **User#friends_and_own_posts** in order to get all posts written by a user on their friends, students can use the simple method in the User model (in order to NOT do it with Enumerable)
```ruby
# User model

def friends_and_own_posts
  Post.where(user: (self.friends << self))
  # This will produce SQL query with IN. Something like: select * from posts where user_id IN (1,45,874,43);
end
```
- [x] **Tests:** There is an issue on 11 of your test, apparently there is an error in importing bootstrap/scss/bootstrap. If you decide to use BootStrap on this project that's great! but make sure you include the proper configuration. I couldn't find it on your Gemfile bootstrap gem and that might be causing the issue. make sure to add a test for each of the functions display above 


# Happy Coding! :nerd_face: :computer:

File: app/models/user.rb [hunk 1]
```diff
@@ -39,7 +27,7 @@ def friend?(user)
     friends.include?(user)
   end
 
-  def friend_requests
-    friendships.map { |friendship| friendship.friend unless friendship.confirmed }.compact
+  def friends_and_own_posts
+    Post.where(user: (friends << self))
   end
 end
```
Comment (blackpintz) [inline app/models/user.rb:33]: Write a test for the above method.

**Review by ghost — CHANGES_REQUESTED** [review]
# Status: Changes Required

### Hello team 😄,

**Your project looks amazing. Great job 💪. There are still a few fixes needed, however, before the project can be approved.**

1). **I will use an example to explain. I have two users, Jen and Ken. They are not friends. When Jen signs in, she can be able to see Ken's posts when she visits Ken's profile. Please fix it so Jen can see Ken's posts, on his profile page, only when they are friends.**

2). **`user.pending_friends` and `user.friend_requests` returns an empty array even though a user has pending friendships and friend_requests.**

3). **Add tests for your `helper` methods. Declare private any methods you are not using outside a module.**

4). **Remember to indicate in your readme setup manual that a user should run `npm install` to download the bootstrap and other necessary dependencies. Otherwise, it throws an error when this command is not run. The previous TSE mentioned it as well.**

5). **Remove the commented code in `spec/helpers/friendships_helper_spec.rb`.**


### Happy coding 🥂.

**PS: You can reach me on slack (Rose Wanjohi).**

**Review by javitocor — APPROVED** [review]
# STATUS APPROVED 🥇 ⭐

Congratulations team! You made it! 👏 
Your project has been approved! 👍 
Remember to fill up the completion form 📑  merge your brand, and move into the next project 🚀 
Good luck! 😃

**Pull request merged**